packages/playwright-core/src/server/ariaSnapshot.ts

```diff
@@ -106,51 +105,155 @@ function applyAttribute(node: AriaTemplateRoleNode, key: string, value: string)
   throw new Error(`Unsupported attribute [${key}] `);
 }
 
-function parseKey(key: string): AriaTemplateNode {
-  const tokenRegex = /\s*([a-z]+|"(?:[^"]*)"|\/(?:[^\/]*)\/|\[.*?\])/g;
-  let match;
-  const tokens = [];
-  while ((match = tokenRegex.exec(key)) !== null)
-    tokens.push(match[1]);
-
-  if (tokens.length === 0)
-    throw new Error(`Invalid key ${key}`);
-
-  const role = tokens[0] as AriaRole;
-
-  let name: string | RegExp = '';
-  let index = 1;
-  if (tokens.length > 1 && (tokens[1].startsWith('"') || tokens[1].startsWith('/'))) {
-    const nameToken = tokens[1];
-    if (nameToken.startsWith('"')) {
-      name = nameToken.slice(1, -1);
-    } else {
-      const pattern = nameToken.slice(1, -1);
-      name = new RegExp(pattern);
-    }
-    index = 2;
-  }
-
-  const result: AriaTemplateRoleNode = { kind: 'role', role, name };
-  for (; index < tokens.length; index++) {
-    const attrToken = tokens[index];
-    if (attrToken.startsWith('[') && attrToken.endsWith(']')) {
-      const attrContent = attrToken.slice(1, -1).trim();
-      const [attrName, attrValue] = attrContent.split('=', 2);
-      const value = attrValue !== undefined ? attrValue.trim() : 'true';
-      applyAttribute(result, attrName, value);
-    } else {
-      throw new Error(`Invalid attribute token ${attrToken} in key ${key}`);
-    }
-  }
-
-  return result;
-}
-
 function normalizeWhitespace(text: string) {
   return text.replace(/[\r\n\s\t]+/g, ' ').trim();
 }
 
 function valueOrRegex(value: string): string | RegExp {
   return value.startsWith('/') && value.endsWith('/') ? new RegExp(value.slice(1, -1)) : normalizeWhitespace(value);
 }
+
+export class KeyParser {
+  private _input: string;
+  private _pos: number;
+  private _length: number;
+
+  static parse(input: string): AriaTemplateNode {
+    return new KeyParser(input)._parse();
+  }
+
+  constructor(input: string) {
+    this._input = input;
+    this._pos = 0;
+    this._length = input.length;
+  }
+
+  private _peek() {
+    return this._input[this._pos] || '';
+  }
+
+  private _next() {
+    if (this._pos < this._length)
+      return this._input[this._pos++];
+    return null;
+  }
+
+  private _eof() {
+    return this._pos >= this._length;
+  }
+
+  private _skipWhitespace() {
+    while (!this._eof() && /\s/.test(this._peek()))
+      this._pos++;
+  }
+
+  private _readIdentifier(): string {
+    if (this._eof())
+      throw new Error('Unexpected end of input when expecting identifier');
+    const start = this._pos;
+    while (!this._eof() && /[a-zA-Z]/.test(this._peek()))
+      this._pos++;
+    return this._input.slice(start, this._pos);
+  }
+
+  private _readString(): string {
+    let result = '';
+    let escaped = false;
+    while (!this._eof()) {
+      const ch = this._next();
+      if (escaped) {
+        result += ch;
```

Comment:
is it fine performance-wise to do by 1 symbol? We could copy slices.

Reply:
This is a template parser, so should not be in a hot code.